### PR TITLE
Move management of root resource state to engine

### PR DESCRIPTION
**Pull request by swgillespie**
This commit fixes a persistent side-by-side issue in the NodeJS SDK by
moving the management of root resource state to the engine. Doing so
adds two new endpoints to the Engine gRPC service: 1) GetRootResource
and 2) SetRootResource, which get and set the root resource
respectively.

Fixes https://github.com/pulumi/pulumi/issues/1476. This is the first attempt to move a little bit of state from the language host to the engine. We've discussed this in passing a few times, both in the context of this issue and other issues such as config. I think that I prefer this particular approach to the problem over the environment variable approach because:

1. Our set of shared environment variables is large already and it is not ideal that we are using process environment variables as shared state, as this is [generally fraught with peril](http://www.club.cc.cmu.edu/~cmccabe/blog_the_setenv_fiasco.html)
2. The engine is a logical place to put language host state, since it is required to be around for the duration of Pulumi program execution
3. Reducing the amount of things that a language host must do is (in my opinion) a win in general, since all code that is written in a particular language's SDK must be duplicated for every language supported by Pulumi

The crux of this PR is that the Engine gRPC endpoint now has Get/SetRootResource as RPC calls, which the language host can use to get and set the root resource state respectively. Today, it is the case that the Stack resource always has an empty provider map, so it is the case that only the root resource's URN is required to be saved; however, in the future we may consider non-empty provider maps for root resources.

For ease of review, this PR is split into two commits: https://github.com/pulumi/pulumi/commit/beae8c67baf860534805a4080cf63d6209004a50, which contains all Protobuf/gRPC-related changes and generated code diffs, and https://github.com/pulumi/pulumi/commit/00c541c2e89693d12ccebd81ffe10845ff502e9f, which contains everything else.

### CODE REVIEW

File: sdk/nodejs/tests/runtime/langhost/run.spec.ts [hunk 1]
```diff
@@ -510,6 +514,41 @@ describe("rpc", () => {
...
+                assert.strictEqual(t, "test:index:MyResource");
+                assert.strictEqual(name, "testResource1");
+                assert.strictEqual(parent, ctx.stackUrn);
+                return { urn: makeUrn(t, name), id: undefined, props: undefined };
+            },
+        },
+        "backcompat_root_resource": {
+            program: path.join(base, "001.one_resource"),
+            expectResourceCount: 2,
+            skipRootResourceEndpoints: true,
```
Comment (swgillespie) [inline sdk/nodejs/tests/runtime/langhost/run.spec.ts]: If it's not clear here, this test deliberately does not provide implementations of `GetRootResource` and `SetRootResource`, which causes the Engine gRPC server to respond with the `UNIMPLEMENTED` status code - which is exactly what would happen if a user used this version of the NodeJS SDK with a Pulumi CLI that doesn't support this RPC endpoint.

**Comment by ellismg** [comment]
> Fixes #1476. 

Just so I am 100% clear on my understanding, this only fixes things when all versions of the pulumi package that are loaded by your program have this fix AND you are using an engine that contains these new RPCs, correct?

**Comment by swgillespie** [comment]
> Just so I am 100% clear on my understanding, this only fixes things when all versions of the pulumi package that are loaded by your program have this fix AND you are using an engine that contains these new RPCs, correct?

Correct. I don't think there's a way we can fix this problem for SDKs that have already been released.

**Comment by swgillespie** [comment]
thanks!

**Pull request merged**